### PR TITLE
feature(lxl): plugin stdim into ppo

### PR DESCRIPTION
**Pull request by lixl-st**
## Check List

- [ ] merge the latest version source branch/repo, and resolve all the conflicts
- [ ] pass style check
- [ ] pass all the tests

### CODE REVIEW

File: dizoo/classic_control/cartpole/config/cartpole_ppo_stdim_config.py [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+            actor_head_hidden_size=128,
+        ),
+        aux_model=dict(
+            encode_shape=64,
+            heads = [1, 1],
+            loss_type = 'infonce',
+            temperature = 1.0,
+        ),
+        # the ratio of the auxiliary loss to the TD loss
+        aux_loss_ratio = 0.003,
```
Comment (puyuan1996) [inline dizoo/classic_control/cartpole/config/cartpole_ppo_stdim_config.py:30]: 不同环境的aux_loss_ratio是手动调的吗？ 变量改为aux_loss_weight可能更合适，ratio表示比值的意思，实际这里应该是aux_loss的权重。

File: ding/policy/ppo.py [hunk 2]
```diff
@@ -954,3 +954,257 @@ def _monitor_vars_learn(self) -> List[str]:
...
+                    data['return'] = unnormalized_return / self._running_mean_std.std
+                    self._running_mean_std.update(unnormalized_return.cpu().numpy())
+                else:
+                    data['return'] = data['adv'] + data['value']
+
+            for batch in split_data_generator(data, self._cfg.learn.batch_size, shuffle=True):
+                # ======================
+                # Auxiliary model update
+                # ======================
+                with torch.no_grad():
```
Comment (puyuan1996) [inline ding/policy/ppo.py:1110]: Auxiliary model与rl model更新的顺序是特别设计的吗？这里是Auxiliary  model先更新，dqn中是Auxiliary  model后更新。应该效果类似，或许可以与dqn保持一致的顺序？

File: ding/policy/ppo.py [hunk 3]
```diff
@@ -954,3 +954,257 @@ def _monitor_vars_learn(self) -> List[str]:
...
+                        ppo_continuous_loss.entropy_loss + ppo_discrete_loss.entropy_loss
+                    )
+                    ppo_info = type(ppo_continuous_info)(
+                        max(ppo_continuous_info.approx_kl, ppo_discrete_info.approx_kl),
+                        max(ppo_continuous_info.clipfrac, ppo_discrete_info.clipfrac)
+                    )
+                # ======================
+                # Compute auxiliary loss
+                # ======================
+                x, y = self._aux_encode(data)
```
Comment (puyuan1996) [inline ding/policy/ppo.py:1165]: 这里是否需要 with torch.no_grad():呢？这里计算的aux_loss_eval 只是用于统计吧，注释可以写一下

File: ding/policy/ppo.py [hunk 4]
```diff
@@ -954,3 +954,257 @@ def _monitor_vars_learn(self) -> List[str]:
...
+                    self._running_mean_std.update(unnormalized_return.cpu().numpy())
+                else:
+                    data['return'] = data['adv'] + data['value']
+
+            for batch in split_data_generator(data, self._cfg.learn.batch_size, shuffle=True):
+                # ======================
+                # Auxiliary model update
+                # ======================
+                with torch.no_grad():
+                    x_no_grad, y_no_grad = self._aux_encode(batch)
```
Comment (puyuan1996) [inline ding/policy/ppo.py:1111]: Auxiliary model的更新和rl更新必须像这样耦合在一起吗？感觉Auxiliary model可以容忍更多的数据重复利用，或许增加Auxiliary model更新次数可以提高学习效率

File: ding/policy/ppo.py (hunk 2 above)
Comment (lixl-st) [reply ding/policy/ppo.py:1110]: 论文中是交替更新的，应该没有严格规定哪个在前。dqn 里面也是 Auxiliary model 先更新的，只是我当时将 Auxiliary model 的 forward 过程写在了 RL model 后面，但真正更新的时机应该以 loss.backward() 为准，这个是 Auxiliary model 在前。为了避免混淆，我把 dqn 那边的顺序调整一下。

File: ding/policy/ppo.py (hunk 3 above)
Comment (lixl-st) [reply ding/policy/ppo.py:1165]: 另外，我想到因为 aux_optimizer 中也没有 Encoding network 的参数，可能在计算 aux_loss_learn 的时候，不需要给 x_encoding 和 y_encoding 加 no_grad 的限制，如果去掉这个限制的话，那可能 aux_loss_learn 和 aux_loss_eval 就完全一样了

File: ding/policy/ppo.py (hunk 4 above)
Comment (lixl-st) [reply ding/policy/ppo.py:1111]: 如果要完全解耦成独立更新的模式的话，可能没法复用现在的 policy 结构，除非能够验证两模型完全独立更新明显优于交替更新的方案。增加 Auxiliary model 的更新次数是一个好主意，但也可能会显著增加耗时，这一点也需要评估一下。

File: dizoo/minigrid/config/minigrid_onppo_stdim_config.py [hunk 5]
```diff
@@ -0,0 +1,69 @@
+from easydict import EasyDict
+
+collector_env_num = 8
+minigrid_ppo_stdim_config = dict(
+    exp_name="minigrid_empty8_onppo_stdim_seed0",
+    env=dict(
+        collector_env_num=8,
```
Comment (puyuan1996) [inline dizoo/minigrid/config/minigrid_onppo_stdim_config.py:7]: `collector_env_num=collector_env_num,`

File: dizoo/minigrid/config/minigrid_onppo_stdim_config.py [hunk 6]
```diff
@@ -0,0 +1,69 @@
+from easydict import EasyDict
+
+collector_env_num = 8
```
Comment (puyuan1996) [inline dizoo/minigrid/config/minigrid_onppo_stdim_config.py:3]: `evaluator_env_num = 5`， 统一在上面设置env_num

File: dizoo/classic_control/cartpole/config/cartpole_ppo_stdim_config.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+        import_names=['dizoo.classic_control.cartpole.envs.cartpole_env'],
+    ),
+    env_manager=dict(type='base'),
+    policy=dict(type='ppo_stdim'),
+)
+cartpole_ppo_stdim_create_config = EasyDict(cartpole_ppo_stdim_create_config)
+create_config = cartpole_ppo_stdim_create_config
+
+if __name__ == "__main__":
+    # or you can enter `ding -m serial_onpolicy -c cartpole_ppo_config.py -s 0`
```
Comment (puyuan1996) [inline dizoo/classic_control/cartpole/config/cartpole_ppo_stdim_config.py:63]: `*cartpole_ppo_stdim_config.py`

File: ding/entry/tests/test_serial_entry_onpolicy.py [hunk 8]
```diff
@@ -65,3 +66,17 @@ def test_onpolicy_ppo_continuous():
...
+def test_onppo_stdim():
+    config = [deepcopy(cartpole_ppo_stdim_config), deepcopy(cartpole_ppo_stdim_create_config)]
+    config[0].policy.learn.update_per_collect = 1
+    config[0].exp_name = 'cartpole_ppo_stdim_unittest'
+    try:
+        serial_pipeline_onpolicy(config, seed=0, max_train_iter=1)
+    except Exception:
+        assert False, "pipeline fail"
+    finally:
+        os.popen('rm -rf cartpole_ppo_stdim_unittest')
```
Comment (puyuan1996) [inline ding/entry/tests/test_serial_entry_onpolicy.py:82]: https://github.com/opendilab/DI-engine/runs/7334248124?check_suite_focus=true unittes还有没有调过的，检查一下

File: ding/policy/ppo.py [hunk 9]
```diff
@@ -954,3 +954,267 @@ def _monitor_vars_learn(self) -> List[str]:
...
+            discount_factor=0.99,
+            # (float) GAE lambda factor for the balance of bias and variance(1-step td and mc)
+            gae_lambda=0.95,
+        ),
+        eval=dict(),
+        aux_loss_weight=0.001,
+    )
+
+    def _init_learn(self) -> None:
+        super()._init_learn()
```
Comment (puyuan1996) [inline ding/policy/ppo.py:1028]: 可以以这里的格式https://github.com/opendilab/DI-engine/blob/main/ding/model/common/head.py，添加详细注释

File: .coveragerc [hunk 10]
```diff
@@ -10,6 +10,7 @@ omit =
   ding/utils/data/tests/test_dataloader.py
   ding/config/utils.py
   ding/entry/tests/test_serial_entry_algo.py
+  ding/entry/tests/test_serial_entry_onpolicy.py
```
Comment (PaParaZz1) [inline .coveragerc:13]: why add this

File: ding/policy/dqn.py [hunk 11]
```diff
@@ -555,6 +557,22 @@ def _forward_learn(self, data: Dict[str, Any]) -> Dict[str, Any]:
...
 
+        # ======================
+        # Auxiliary model update
+        # ======================
+        # RL network encoding
+        # To train the auxiliary network, the gradients of x, y should be 0.
+        with torch.no_grad():
+            x_no_grad, y_no_grad = self._model_encode(data)
+        # the forward function of the auxiliary network
+        aux_loss_learn = self._aux_model.forward(x_no_grad, y_no_grad)
```
Comment (PaParaZz1) [inline ding/policy/dqn.py:568]: add `aux_loss_learn` in tb

File: ding/policy/dqn.py [hunk 12]
```diff
@@ -519,11 +519,13 @@ def _get_encoding_size(self):
...
+        if self._cuda:
+            test_data = to_device(test_data, self._device)
         with torch.no_grad():
-            x, y = self._aux_encode(test_data)
+            x, y = self._model_encode(test_data)
         return x.size()[1:], y.size()[1:]
 
-    def _aux_encode(self, data):
+    def _model_encode(self, data):
         x = self._model.encoder(data["obs"])
```
Comment (PaParaZz1) [inline ding/policy/dqn.py:529]: add a assertion to give more information for model definition

File: ding/policy/ppo.py [hunk 13]
```diff
@@ -954,3 +954,277 @@ def _monitor_vars_learn(self) -> List[str]:
...
+        """
+        super()._init_learn()
+        x_size, y_size = self._get_encoding_size()
+        self._aux_model = ContrastiveLoss(x_size, y_size, **self._cfg.aux_model)
+        if self._cuda:
+            self._aux_model.cuda()
+        self._aux_optimizer = Adam(self._aux_model.parameters(), lr=self._cfg.learn.learning_rate)
+        self._aux_loss_weight = self._cfg.aux_loss_weight
+
+    def _get_encoding_size(self):
```
Comment (PaParaZz1) [inline ding/policy/ppo.py:1051]: maybe we can abstract a general function and place it into `ding/torch_utils`

File: ding/policy/ppo.py [hunk 14]
```diff
@@ -954,3 +954,277 @@ def _monitor_vars_learn(self) -> List[str]:
...
+        """
+        Overview:
+            Get the input encoding size of the ST-DIM axuiliary model.
+        Returns:
+            - info_dict (:obj:`Tuple[Tuple]`): the encoding size without the first (Batch) dimension.
```
Comment (PaParaZz1) [inline ding/policy/ppo.py:1056]: - `Tuple[Tuple, Tuple]`?
- use uppercase letter for comment, i.e., `The encoding size`

**Review by PaParaZz1 — CHANGES_REQUESTED** [review]

File: ding/policy/ppo.py (hunk 13 above)
Comment (lixl-st) [reply ding/policy/ppo.py:1051]: I once thought about this, but we need to define the network and fake input for each scene in advance. The function may be just like this:

> def func(network, data):
>  \ return netwotk(data).size()

I think it is not a very good abbreviation

File: ding/policy/dqn.py (hunk 12 above)
Comment (lixl-st) [reply ding/policy/dqn.py:529]: done

File: ding/policy/dqn.py (hunk 11 above)
Comment (lixl-st) [reply ding/policy/dqn.py:569]: done

**Review by PaParaZz1 — APPROVED** [review]
